cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -344,36 +303,80 @@ func performEtcdStaticPodUpgrade(client clientset.Interface, waiter apiclient.Wa
+	if strings.Compare(desiredEtcdVersion.String(), currentEtcdVersion.String()) == 0 {
+		return false, nil
+	}
+	return true, nil
+}
+
+// rollbackEtcdOnFailedUpgrade is used to rolled back oldEtcd
+func rollbackEtcdOnFailedUpgrade(pathMgr StaticPodPathManager, cfg *kubeadmapi.InitConfiguration, recoverManifests map[string]string, oldEtcdClient etcdutil.ClusterInterrogator, backupEtcdDir string, err error) error {
+	// Despite the fact that upgradeComponent was successful, there is something wrong with the etcd cluster
+	// First step is to restore back up of datastore
+	fmt.Println("[upgrade/etcd] Rolling back etcd data")
+	if err := rollbackEtcdData(cfg, pathMgr); err != nil {
+		// Even copying back datastore failed, no options for recovery left, bailing out
+		return errors.Wrapf(err, "fatal error rolling back local etcd cluster datadir, the backup of etcd database is stored here:(%s)", backupEtcdDir)
+	}
+	fmt.Println("[upgrade/etcd] Etcd data rollback successful")
+
+	if recoverManifests != nil {
 		// Old datastore has been copied, rolling back old manifests
 		fmt.Println("[upgrade/etcd] Rolling back etcd manifest")
 		rollbackOldManifests(recoverManifests, err, pathMgr, true)
 		// rollbackOldManifests() always returns an error -- ignore it and continue
+	}
 
-		// Assuming rollback of the old etcd manifest was successful, check the status of etcd cluster again
-		fmt.Println("[upgrade/etcd] Waiting for previous etcd to become available")
-		if _, err := oldEtcdClient.WaitForClusterAvailable(retries, retryInterval); err != nil {
-			fmt.Printf("[upgrade/etcd] Failed to healthcheck previous etcd: %v\n", err)
-			// Nothing else left to try to recover etcd cluster
-			return true, errors.Wrapf(err, "fatal error rolling back local etcd cluster manifest, the backup of etcd database is stored here:(%s)", backupEtcdDir)
-		}
-		fmt.Println("[upgrade/etcd] Etcd was rolled back and is now available")
-
-		// We've successfully rolled back etcd, and now return an error describing that the upgrade failed
-		return true, errors.Wrap(err, "fatal error upgrading local etcd cluster, rolled the state back to pre-upgrade state")
+	// Assuming rollback of the old etcd manifest was successful, check the status of etcd cluster again
+	fmt.Println("[upgrade/etcd] Waiting for previous etcd to become available")
+	if _, err := oldEtcdClient.WaitForClusterAvailable(etcdConnectRetries, etcdConnectRetryInterval); err != nil {
+		fmt.Printf("[upgrade/etcd] Failed to healthcheck previous etcd: %v\n", err)
+		// Nothing else left to try to recover etcd cluster
+		return errors.Wrapf(err, "fatal error rolling back local etcd cluster manifest, the backup of etcd database is stored here:(%s)", backupEtcdDir)
 	}
+	fmt.Println("[upgrade/etcd] Etcd was rolled back and is now available")
 
-	return false, nil
+	// We've successfully rolled back etcd, and now return an error describing that the upgrade failed
```

Comment:
This comment needs to be moved to the caller function too. It goes along with the returned error there.